### ISSUE TITLE
Different result between `flash_attn_varlen_kvpacked_func` and `torch.scaled_dot_product_attention`

### ISSUE THREAD

**Issue by YJYJLee** [Dao-AILab/flash-attention#581]
Hi, I am testing numerical differences between `flash_attn_varlen_kvpacked_func` and `torch.scaled_dot_product_attention`.

When I don't use causal mask, there is almost no numerical differences (average 0.0003) between `flash_attn_varlen_kvpacked_func` and `torch.scaled_dot_product_attention`. 
![Screenshot 2023-09-29 at 11 49 18 AM](https://github.com/Dao-AILab/flash-attention/assets/28900943/bc94f455-e97c-4b0c-9ddd-1e24bbd239af)

However, when I enable causal mask, then I have a huge numerical differences (average 0.24). Some of the outputs of `flash_attn_varlen_kvpacked_func` is 0.

![Screenshot 2023-09-29 at 11 50 35 AM](https://github.com/Dao-AILab/flash-attention/assets/28900943/b18429c4-2099-4088-bf5a-a9a8ca4430f3)

The code I used for causal mask testing is as below
```
// torch.scaled_dot_product_attention
with sdp_kernel(**backend_map[SDPBackend.MATH]):
    torch_result = F.scaled_dot_product_attention(query, key, value, attn_mask=causal_key_padding_mask)


// flash_attn_varlen_kvpacked_func
output_pad_fn = lambda output_unpad: rearrange(output_unpad, "(b s) h d -> b s h d", b=query.shape[0])
new_q = query.transpose(1, 2)
new_k = key.transpose(1, 2)
new_v = value.transpose(1, 2)
q_unpad = rearrange(new_q, "b s h d -> (b s) h d")
max_seqlen_q = new_q.shape[1]
cu_seqlens_q = torch.arange(0, (new_q.shape[0] + 1) * max_seqlen_q, step=max_seqlen_q, dtype=torch.int32, device=device)
k_unpad, indices_k, cu_seqlens_k, max_seqlen_k = unpad_input(new_k, binary_key_padding_mask)
v_unpad, _, _, _ = unpad_input(new_v, binary_key_padding_mask)
kv_unpad = torch.stack([k_unpad, v_unpad], dim=1)
out_unpad = flash_attn_varlen_kvpacked_func(q_unpad, kv_unpad, cu_seqlens_q, cu_seqlens_k, max_seqlen_q, max_seqlen_k, causal=True)
out = output_pad_fn(out_unpad)
```

I created `causal_key_padding_mask` as below
```
causal_key_padding_mask = torch.full([batch_size, num_heads, max_sequence_len, max_sequence_len], 1, device=device, dtype=dtype, requires_grad=False)

for i in range(1, max_sequence_len):
    causal_key_padding_mask[:,:,i-1,i:] = 0
causal_key_padding_mask += key_padding_mask
```

to test w/o causal mask, I used `key_padding_mask` for `torch.scaled_dot_product_attention` as attn_mask instead, and set `causal=False` for `flash_attn_varlen_kvpacked_func`

**Comment by tridao** [comment]
As mentioned in the [README](https://github.com/Dao-AILab/flash-attention/tree/main#changes-in-v21-compared-to-v20), the behavior with `causal` when seqlen_q != seqlen_k has changed since FA v2.1.
F.sdpa is still using the old behavior.

**Comment by drisspg** [comment]
Some more context here: https://github.com/pytorch/pytorch/issues/108108

**Comment by YJYJLee** [comment]
@tridao @drisspg  Thank you so much for the info!!

**Comment by jiangwangyi** [comment]
It seems that `xformers.ops.LowerTriangularMask` is also using the old behavior?

**Comment by jiangwangyi** [comment]
> As mentioned in the [README](https://github.com/Dao-AILab/flash-attention/tree/main#changes-in-v21-compared-to-v20), the behavior with `causal` when seqlen_q != seqlen_k has changed since FA v2.1. F.sdpa is still using the old behavior.

It seems that `xformers.ops.LowerTriangularMask` is also using the old behavior?

**Comment by drisspg** [comment]
For PyTorch SDPA I currently have a PR that will enable lower_right causal masking: https://github.com/pytorch/pytorch/pull/113103/files#diff-7cb9608c2ed3ac9f01835fd522391e5aa048170bd0754b2ec9ec3380cc0889fbR156
For the representative API
